gen/field.go

```diff
@@ -94,23 +105,88 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tk := []string{jsonTagName} // list of tag keys to preserve order
+	ts := map[string]string{
+		jsonTagName: compileJSONTagValue(f, f.Name),
+	}
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.Annotations[goTagName]; goAnnotation != "" {
+		goTags, goTagKeys := mapFromRawTag(goAnnotation)
+		for _, k := range goTagKeys {
+			if _, exists := ts[k]; !exists {
+				tk = append(tk, k)
+			}
+			ts[k] = goTags[k]
+		}
+
+		if jt := reflect.StructTag(goAnnotation).Get(jsonTagName); jt != "" {
+			ts[jsonTagName] = compileJSONTagValue(f, jt)
+		}
+	}
+	// TODO (abg): Take js.name annotation into account
+	return mapToRawTag(ts, tk)
+}
+
+// mapFromRawTag constructs a tag's key, value map from the raw tag string
+// `json:"foo,required" bson:"foo bar"` -> {"json": "foo,required", "bson": "foo bar"}
+// it also returns a list of keys to preserve tag order
+func mapFromRawTag(a string) (map[string]string, []string) {
+	ts, tk := map[string]string{}, []string{}
+	split := strings.Split(a, " ")
+	for i, t := range split {
+		if t[len(t)-1] != '"' {
+			split[i+1] = strings.Join(split[i:i+2], " ")
+			continue
+		}
+		k := strings.Split(t, ":")[0]
```

Comment:
This is perhaps not the safest way to implement this. I'm guessing there are corner cases you or I haven't thought of. Escaped quotes (`\"` come to mind).

Unless we're willing to replicate the logic from reflect.StructTag, we should maybe use https://github.com/fatih/structtag for this. That library seems more battle tested because it's used for vim-go's struct tag integration.

Note from the readme,

> Please vendor the library as it might change in future versions.

So you'll want to put it in vendor/ and check it in. To do this cleanly, `glide get` the library as usual, remove vendor/ from .gitignore, and then in vendor/ add a .gitignore with the following contents:

```
# We will explicitly whitelist vendored libraries that we want to check in.
/*

!/github.com/fatih/structtag
```